File: nbody/java/NBodyPerf.java
import java.util.Random;

public class NBodyPerf {

    static final int N = 10000;          // Number of bodies
    static final int STEPS = 10;         // Number of simulation steps
    static final double DT = 0.01;       // Time step
    static final double G = 6.67430e-11; // Gravitational constant

    static class Body {
        double x, y, z;
        double vx, vy, vz;
        double mass;
    }

    static Body[] initBodies(int n) {
        Random rand = new Random();
        Body[] bodies = new Body[n];
        for (int i = 0; i < n; i++) {
            Body b = new Body();
            b.x = rand.nextDouble() * 1e5;
            b.y = rand.nextDouble() * 1e5;
            b.z = rand.nextDouble() * 1e5;
            b.vx = b.vy = b.vz = 0.0;
            b.mass = rand.nextDouble() * 1e22 + 1e20;
            bodies[i] = b;
        }
        return bodies;
    }

    static void simulate(Body[] bodies, int steps, double dt) {
        for (int s = 0; s < steps; s++) {
            double[] fx = new double[N];
            double[] fy = new double[N];
            double[] fz = new double[N];

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (i == j) continue;
                    double dx = bodies[j].x - bodies[i].x;
                    double dy = bodies[j].y - bodies[i].y;
                    double dz = bodies[j].z - bodies[i].z;
                    double dist = Math.sqrt(dx * dx + dy * dy + dz * dz) + 1e-9;
                    double force = G * bodies[i].mass * bodies[j].mass / (dist * dist);
                    fx[i] += force * dx / dist;
                    fy[i] += force * dy / dist;
                    fz[i] += force * dz / dist;
                }
            }

            for (int i = 0; i < N; i++) {
                bodies[i].vx += fx[i] / bodies[i].mass * dt;
                bodies[i].vy += fy[i] / bodies[i].mass * dt;
                bodies[i].vz += fz[i] / bodies[i].mass * dt;
            }

            for (int i = 0; i < N; i++) {
                bodies[i].x += bodies[i].vx * dt;
                bodies[i].y += bodies[i].vy * dt;
                bodies[i].z += bodies[i].vz * dt;
            }
        }
    }

    public static void main(String[] args) {
        System.out.printf("Running N-body simulation with %d bodies for %d steps...\n", N, STEPS);
        System.out.println("Technology: Java");

        Body[] bodies = initBodies(N);

        long start = System.nanoTime();
        simulate(bodies, STEPS, DT);
        long end = System.nanoTime();

        double elapsed = (end - start) / 1e9;
        System.out.printf("%d-bodies for %d steps: %.6f seconds\n", N, STEPS, elapsed);
    }
}

File: nbody/go/nbody_perf.go
package main

import (
    "fmt"
    "math"
    "math/rand"
    "time"
)

const (
    N     = 10000         // Number of bodies
    STEPS = 10            // Number of simulation steps
    DT    = 0.01          // Time step
    G     = 6.67430e-11   // Gravitational constant
)

type Body struct {
    x, y, z   float64
    vx, vy, vz float64
    mass      float64
}

func initBodies(n int) []Body {
    bodies := make([]Body, n)
    for i := range bodies {
        bodies[i] = Body{
            x:    rand.Float64() * 1e5,
            y:    rand.Float64() * 1e5,
            z:    rand.Float64() * 1e5,
            vx:   0.0,
            vy:   0.0,
            vz:   0.0,
            mass: rand.Float64()*1e22 + 1e20,
        }
    }
    return bodies
}

func simulate(bodies []Body, steps int, dt float64) {
    for s := 0; s < steps; s++ {
        for i := range bodies {
            fx, fy, fz := 0.0, 0.0, 0.0
            for j := range bodies {
                if i == j {
                    continue
                }
                dx := bodies[j].x - bodies[i].x
                dy := bodies[j].y - bodies[i].y
                dz := bodies[j].z - bodies[i].z
                dist := math.Sqrt(dx*dx + dy*dy + dz*dz) + 1e-9
                force := G * bodies[i].mass * bodies[j].mass / (dist * dist)
                fx += force * dx / dist
                fy += force * dy / dist
                fz += force * dz / dist
            }
            bodies[i].vx += fx / bodies[i].mass * dt
            bodies[i].vy += fy / bodies[i].mass * dt
            bodies[i].vz += fz / bodies[i].mass * dt
        }
        for i := range bodies {
            bodies[i].x += bodies[i].vx * dt
            bodies[i].y += bodies[i].vy * dt
            bodies[i].z += bodies[i].vz * dt
        }
    }
}

func main() {
    fmt.Printf("Running N-body simulation with %d bodies for %d steps...\n", N, STEPS)
    fmt.Println("Technology: Go")

    rand.Seed(time.Now().UnixNano())
    bodies := initBodies(N)

    start := time.Now()
    simulate(bodies, STEPS, DT)
    elapsed := time.Since(start)

    fmt.Printf("%d-bodies for %d steps: %.6f seconds\n", N, STEPS, elapsed.Seconds())
}

